plugins/outputs/zabbix/lld.go

```diff
@@ -0,0 +1,196 @@
+		setValues := make(map[string]string, len(set))
+		for tagKey, tagValue := range set {
+			if tagKey == "host" {
+				continue
+			}
+			setValues[fmt.Sprintf("{#%s}", strings.ToUpper(tagKey))] = tagValue
+		}
+		values = append(values, setValues)
+	}
+
+	return map[string][]map[string]string{"data": values}
+}
+
+// hostLLD used as a key to identify different LLDs in the same, or different, hosts
+type hostLLD struct {
+	Hostname string
+	LLDKey   string
+}
+
+// Push returns a slice of metrics to send to Zabbix with the LLD data using the accumulated info.
+// The name of the metric will be always "lld" (const).
+// It will have only one tag, with the host.
+// It will have an uniq field, with the LLD key as the key name and the JSON data as the value
+// Eg.: lld,host=hostA disk.device.fstype.mode.path="{\"data\":[...
+func (zl *zabbixLLD) Push() []telegraf.Metric {
+	metrics := make([]telegraf.Metric, 0, len(zl.receivedData))
+
+	// Iterate over the data collected in the last period.
+	// Compare with the data sent the last time.
+	// If different, send a new LLD.
+	for key, tags := range zl.receivedData {
+		// Skip already sent LLDs
+		if zl.compareAndDelete(key, tags) {
+			continue
+		}
+
+		dataValues := tags.generateDataValues()
+
+		dataValuesJSON, err := json.Marshal(dataValues)
+		if err != nil {
+			zl.log.Warnf("Marshaling to JSON LLD tags to Zabbix format: %v", err)
+		}
+
+		// Add the LLD m
+		m := metric.New(
+			lldName,
+			map[string]string{
+				"host": key.Hostname,
+			},
+			map[string]interface{}{
+				key.LLDKey: dataValuesJSON,
+			},
+			time.Now(),
+		)
+		metrics = append(metrics, m)
+	}
+
+	// Send empty LLDs for the LLDs that were sent the last time but not this time.
+	for key := range zl.previousReceivedData {
+		emptyDataValuesJSON, err := json.Marshal(map[string]interface{}{"data": []interface{}{}})
+		if err != nil {
+			zl.log.Warnf("Marshaling to JSON empty data Zabbix format: %v", err)
+		}
+
+		// Add the LLD m
+		m := metric.New(
+			lldName,
+			map[string]string{
+				"host": key.Hostname,
+			},
+			map[string]interface{}{
+				key.LLDKey: emptyDataValuesJSON,
+			},
+			time.Now(),
+		)
+		metrics = append(metrics, m)
+	}
+
+	// Move current data to previousReceivedData
+	zl.previousReceivedData = zl.receivedData
+
+	// Reset receivedData and cache to receive data from the new period
+	zl.receivedData = make(map[hostLLD]lldTags, len(zl.receivedData))
+	zl.cache = make(map[uint64]bool, len(zl.cache))
+
+	// Clear previousReceivedData if enough time has passed
+	if time.Since(zl.lastClear) > time.Duration(zl.clearInterval) {
+		zl.previousReceivedData = make(map[hostLLD]lldTags, len(zl.previousReceivedData))
+		zl.lastClear = time.Now()
+	}
+
+	return metrics
+}
+
+// Add parse a metric and add it to the LLD cache.
+func (zl *zabbixLLD) Add(in telegraf.Metric) {
+	hostname, err := getHostname(in.TagList())
+	if err != nil {
+		zl.log.Errorf("Error getting hostname: %v", err)
+		return
+	}
```

Comment:
How about returning an error in this function and log the result in the caller?

Reply:
Fixed in 06d2d9053